clients/core/src/main/scala/mqperf/Sender.scala

```diff
@@ -21,47 +24,63 @@ class Sender(config: Config, mq: Mq, clock: Clock) extends StrictLogging {
 
   def run(): Future[Unit] = {
     val mqSender = mq.createSender(config)
+    val end = clock.millis() + config.testLengthSeconds.seconds.toMillis
+    runIterations(mqSender, end)
+      .map { _ =>
+        logger.info("Sending done, waiting for all messages to be flushed ...")
+        Thread.sleep(3.seconds.toMillis)
+      }
+      .flatMap(_ => mqSender.close())
+  }
 
-    Future {
-      blocking {
-        val start = clock.millis()
-        val end = start + config.testLengthSeconds * 1000L
-
-        val permits = new AtomicInteger(config.maxSendInFlight)
-        val batchSize = config.batchSizeSend
+  private def runIterations(mqSender: MqSender, testEnd: Long): Future[Unit] = {
+    val batchSize = config.batchSizeSend
+    val sentMessages = new AtomicInteger(0)
+    val permits = new AtomicInteger(0)
 
-        while (clock.millis() < end) {
-          val iterationStart = clock.millis()
-          // we send at most `config.msgsPerSecond` messages, but no more than the number of available permits
-          val msgsToSend = math.min(config.msgsPerSecond, permits.getAndUpdate(p => math.max(0, p - config.msgsPerSecond)))
-          logger.info(s"Messages to send: $msgsToSend")
-          val batches = msgsToSend / batchSize
-          (1 to batches).foreach { _ =>
-            val sendStart = clock.millis()
-            mqSender.send(nextMessagesBatch()).onComplete { result =>
-              permits.addAndGet(batchSize)
-              LocalMetrics.messageCounter.inc(batchSize.toDouble)
-              LocalMetrics.messageLatencyHistogram.observe((clock.millis() - sendStart).toDouble)
-              result match {
-                case Failure(t) => logger.error("Exception when sending a batch of messages", t)
-                case Success(_) =>
-              }
-            }
+    def send(): Future[Unit] = {
+      if (clock.millis() >= testEnd) {
+        Future.successful(())
+      } else if (sentMessages.get() < permits.get()) {
+        val sendStart = clock.millis()
+        sentMessages.addAndGet(batchSize)
+        mqSender
+          .send(nextMessagesBatch(batchSize))
+          .map { _ =>
+            LocalMetrics.messageCounter.inc(batchSize.toDouble)
+            LocalMetrics.messageLatencyHistogram.observe((clock.millis() - sendStart).toDouble)
           }
+          .flatMap(_ => send())
+      } else {
+        // do nothing, just wait for either: increasing of the permits or the test end
+        send()
```

Comment:
this will busy-loop the CPU, if there are not enough permits, until they arrive

Reply:
would be better to instead somehow wait until enough permits are available

Reply:
I used `Deferred` from cats  to wait in non-blocking manner.